system-addon/lib/SearchFeed.jsm

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+ /* globals ContentSearch, XPCOMUtils, windowMediator, Services */
+"use strict";
+
+const {utils: Cu} = Components;
+const {actionTypes: at, actionCreators: ac} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+const SEARCH_ENGINE_TOPIC = "browser-search-engine-modified";
+
+Cu.import("resource:///modules/ContentSearch.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+
+XPCOMUtils.defineLazyServiceGetter(this, "windowMediator",
+  "@mozilla.org/appshell/window-mediator;1",
+  "nsIWindowMediator");
+
+this.SearchFeed = class SearchFeed {
+  getCurrentBrowser() {
+    const win = windowMediator.getMostRecentWindow("navigator:browser");
+    const browser = win.getBrowser().selectedBrowser;
+    return browser;
+  }
+  addObservers() {
+    Services.obs.addObserver(this, SEARCH_ENGINE_TOPIC, false);
+  }
+  observe(subject, topic, data) {
+    if (topic !== SEARCH_ENGINE_TOPIC) {
+      return;
+    }
+    switch (data) {
+      case "engine-current":
+      case "engine-changed":
+      case "engine-added":
+      case "engine-removed":
+        this.getState();
+        break;
+    }
+  }
+  async getState() {
+    const state = await ContentSearch.currentStateObj(true);
+    const action = {type: at.SEARCH_STATE_UPDATED, data: {engines: state.engines, currentEngine: state.currentEngine}};
+    this.store.dispatch(ac.BroadcastToContent(action));
+  }
+  performSearch(action) {
+    let gBrowser = this.getCurrentBrowser();
+    ContentSearch.performSearch({target: gBrowser}, action.data);
```

Comment:
Yup, can do. I just have to change the message manager  to send up the entire message instead of just msg.data: https://github.com/mozilla/activity-stream/blob/master/system-addon/lib/MessageManager.jsm#L145
That way I can do msg.target.browser, and I can access the browser it came from